src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -138,12 +140,15 @@ public final void reset(Collection<? extends Whitelist> delegates) {
                     this.delegates.add(delegate);
                 }
             }
-            for (ProxyWhitelist pw : wrappers.keySet()) {
-                pw.reset();
+            Iterator<ProxyWhitelist> pwIterator = wrappers.keySet().iterator();
+            synchronized (wrappers) {
```

Comment:
So this is now inside `lock.writeLock()`, and the wrappers will in turn acquire their own r/w lock and then monitor. Seems risky. I guess there should not be a deadlock so long as there are no cycles in `wrappers`, which logically there should not be…

Reply:
That's my understanding as well that as long as there is no cycle (and I can't find a way to produce one) it should be fine.